ui/src/util/auth/provider.tsx

```diff
@@ -0,0 +1,186 @@
+
+  //       if (!res.ok) {
+  //         setLoading(false);
+  //         throw new Error("Logout failed");
+  //       }
+
+  //       return res.json();
+  //     },
+  //     onSuccess: () => {
+  //       setToken(null);
+  //       setUser(null);
+  //       queryClient.removeQueries();
+  //       queryClient.invalidateQueries();
+  //       setLoading(false);
+  //     },
+  //   });
+
+  async function fetchWithoutAuth(path: string, init?: RequestInit) {
+    const res = await fetch(`${API_URL}${path}`, {
+      ...init,
+      headers: {
+        ...(init?.headers || {}),
+      },
+    });
+
+    return res;
+  }
+
+  async function fetchWithAuth(
+    path: string,
+    init?: RequestInit,
+    overrideToken?: string
+  ) {
+    let res;
+    if (!overrideToken && !token) {
+      res = await fetchWithoutAuth(path, init);
+    } else {
+      res = await fetch(`${API_URL}${path}`, {
+        ...init,
+        headers: {
+          ...(init?.headers || {}),
+          Authorization: `Bearer ${overrideToken ?? token}`,
+        },
+      });
+
+      if (res.status === 401) {
+        // This means the token has expired, so remove it
+        alert("You've been logged out!");
+        setUser(null);
+        setToken(null);
+        localStorage.removeItem("jwt");
+        window.postMessage({ command: "parakeet-removeAuthToken" }, origin);
+        queryClient.setQueryData(["auth", "user"], null);
+        res = fetchWithoutAuth(path, init);
+      }
+    }
+    return res;
+  }
+
+  // Get the current user on page load
+  const userQuery = useQuery({
+    queryKey: ["auth", "user"],
+    queryFn: async () => {
+      const token = localStorage.getItem("jwt");
+      if (!token) {
+        return null;
+      }
+      setToken(token);
+      window.postMessage({ command: "parakeet-setAuthToken", token }, origin);
+
+      const res = await fetchWithAuth(
+        "/users/show",
+        undefined,
+        token ?? undefined
+      );
+
+      // TODO: Handle if the server is down
+      const json = await res.json();
+      if (!json.username) {
+        setUser(null);
+      } else {
+        setUser(json);
+      }
+      return json;
+    },
+    staleTime: Infinity,
+    refetchOnWindowFocus: false,
+  });
+
+  const login = async (creds: { email: string; password: string }) => {
+    await loginMut.mutateAsync(creds);
+  };
+
+  const logout = async () => {
+    // Enable these when we have tokens that last more than 30 mins
+    // setLoading(true);
+    // await logoutMut.mutateAsync();
+    // return success;
+
+    setUser(null);
+    localStorage.removeItem("jwt");
+    window.postMessage({ command: "parakeet-removeAuthToken" }, origin);
+    queryClient.setQueryData(["auth", "user"], null);
+    window.location.reload();
+  };
+
+  const value: AuthContextValue = {
+    user,
+    token,
+    userLoading: userQuery.isLoading,
+    loginLoading: loginMut.isPending,
+    login,
+    logout,
+    fetchWithAuth,
```

Comment:
consider wrapping these context functions, like login, logout, and fetchWithAuth, with `React.useCallback` for better performance!